src/java.base/share/classes/java/lang/StringTemplate.java

```diff
@@ -0,0 +1,692 @@
+     * SimpleProcessor<String> processor = st -> {
+     *     List<String> fragments = st.fragments();
+     *     List<Object> values = st.values();
+     *     // check or manipulate the fragments and/or values
+     *     ...
+     *     return StringTemplate.interpolate(fragments, values);
+     * };
+     * }
+     * The {@link FunctionalInterface} {@link StringProcessor} is supplied if
+     * the processor returns {@link String};
+     * {@snippet :
+     * StringProcessor processor = st -> {
+     *     List<String> fragments = st.fragments();
+     *     List<Object> values = st.values();
+     *     // check or manipulate the fragments and/or values
+     *     ...
+     *     return StringTemplate.interpolate(fragments, values);
+     * };
+     * }
+     * The {@link StringTemplate#interpolate()} method is available for those processors
+     * that just need to work with the string interpolation;
+     * {@snippet :
+     * StringProcessor processor = StringTemplate::interpolate;
+     * }
+     * or simply transform the string interpolation into something other than
+     * {@link String};
+     * {@snippet :
+     * SimpleProcessor<JSONObject> jsonProcessor = st -> new JSONObject(st.interpolate());
+     * }
+     * @implNote The Java compiler automatically imports {@link StringTemplate#STR}
+     *
+     * @param <R>  Processor's process result type
+     * @param <E>  Exception thrown type
+     *
+     * @see SimpleProcessor
+     * @see StringProcessor
+     * @see StringTemplate
+     * @see java.util.FormatProcessor
+     *
+     * @since 21
+     *
+     * @jls 15.8.6 Process Template Expressions
+     */
+    @PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+    @FunctionalInterface
+    public interface Processor<R, E extends Throwable> {
+
+        /**
+         * Constructs a result based on the template fragments and values in the
+         * supplied {@link StringTemplate stringTemplate} object.
+         *
+         * @param stringTemplate  a {@link StringTemplate} instance
+         *
+         * @return constructed object of type R
+         *
+         * @throws E exception thrown by the template processor when validation fails
+         */
+        R process(StringTemplate stringTemplate) throws E;
+
+        /**
+         * Built-in policies using this additional interface have the flexibility to
+         * specialize the composition of the templated string by returning a customized
+         * {@link MethodHandle} from {@link Linkage#linkage linkage}.
+         * These specializations are typically implemented to improve performance;
+         * specializing value types or avoiding boxing and vararg arrays.
+         *
+         * @implNote This interface is sealed to only allow standard processors.
+         *
+         * @since 21
+         */
+        @PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+        public sealed interface Linkage permits FormatProcessor {
```

Comment:
I fail to see any reason a user would call `linkage`; is there anything that's not covered by `TemplateRuntime::processStringTemplate`? And if users can't use this, this can just be relocated to one of the `jdk.internal` packages.

Reply:
Linkage is unfinished business for the first preview. What we are working on is a managed construction of the BSM, a ProcessorFactory per se. The hope is that this will be flushed out for the next round.